### ISSUE TITLE
Add a webui for Giraffa

### ISSUE THREAD

**Issue by shvachko** [GiraffaFS/giraffa#30]
Original [issue 30](https://code.google.com/p/giraffa/issues/detail?id=30) created by shvachko on 2012-09-14T20:41:59.000Z:

We can add a webui the same way NameNode implements a webui simply by using org.apache.hadoop.httpServer.

**Comment by shvachko** [comment]
Comment [#1](https://code.google.com/p/giraffa/issues/detail?id=30#c1) originally posted by shvachko on 2012-10-23T07:53:25.000Z:

This is much harder than it seems. I gave it a shot but there is so much to do in terms of JSP and servlets. Someday I'd love to see a nice webUI though.

**Comment by shvachko** [comment]
Comment [#2](https://code.google.com/p/giraffa/issues/detail?id=30#c2) originally posted by shvachko on 2012-10-23T07:53:37.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#3](https://code.google.com/p/giraffa/issues/detail?id=30#c3) originally posted by shvachko on 2013-02-15T18:24:40.000Z:

I think I can work on this one. I've got good ideas on how to embed UI and make fast and lightweight. Any ideas of what needs to be shown in the UI? Any other comments or suggestions?

**Comment by shvachko** [comment]
Comment [#4](https://code.google.com/p/giraffa/issues/detail?id=30#c4) originally posted by shvachko on 2013-02-15T18:36:47.000Z:

Kosta, you may want to hold off on this one. This is actually an interesting issue to tackle because we aren't sure what level to work at. We can't really modify the NameNode webUI because we will replace NameNode basically... I think before this issue we actually need to discuss where the webserver is located on Giraffa.

Will it be on every HBase master? That's my best guess right now. But that will require extensive modifications because it will need to be able to access the FileSystem too.

Perhaps this issue should be changed to address HOW we plan to incorporate a webserver and a webUI for Giraffa, rather than trying to add one in right now.

**Comment by shvachko** [comment]
Comment [#5](https://code.google.com/p/giraffa/issues/detail?id=30#c5) originally posted by shvachko on 2013-02-19T21:06:47.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#6](https://code.google.com/p/giraffa/issues/detail?id=30#c6) originally posted by shvachko on 2013-02-20T07:59:12.000Z:

The first version of webUI should be fairly simple. It could be a simple jsp saying this is Giraffa, contain build info and links to HBase and NameNode as a block manager.
Its the start and we can add information as we go.
Kostya what were your ideas?

**Comment by shvachko** [comment]
Comment [#7](https://code.google.com/p/giraffa/issues/detail?id=30#c7) originally posted by shvachko on 2013-02-20T08:00:34.000Z:

I have no doubt the webUI can be done. My question is WHO / WHAT will host the web server?

**Comment by shvachko** [comment]
Comment [#8](https://code.google.com/p/giraffa/issues/detail?id=30#c8) originally posted by shvachko on 2013-02-20T18:55:37.000Z:

Giraffa jsp should live on HBase. HBase's web server will host it. Say you go to
hbase.master.com:60010/giraffa
and you see the page

**Comment by shvachko** [comment]
Comment [#9](https://code.google.com/p/giraffa/issues/detail?id=30#c9) originally posted by shvachko on 2013-02-20T19:07:23.000Z:

There we go. This will work for now -- just provide a link to NameNode webUIs, etc. Later though we will need to discuss how we are going to access the GiraffaFileSystem from this JSP. That will require a sort of merge with NameNode's current webUI.

I would love to see a simple Giraffa JSP page though on an HBase master. :)

**Comment by shvachko** [comment]
Comment [#10](https://code.google.com/p/giraffa/issues/detail?id=30#c10) originally posted by shvachko on 2013-02-20T20:11:54.000Z:

One suggestion that I have is to use embedded http server e.g. http://elonen.iki.fi/code/nanohttpd/  which doesn't require to bring in tons of jsp-servlet-related dependencies. Giraffa webpage will be mostly static and have minimal number of pages. I have an examplae  of a very lightweight framework which supports simple templates and can serve pages. I'll do a sample page and publish as a patch. then we can evaluate if it's good option.

**Comment by shvachko** [comment]
Comment [#12](https://code.google.com/p/giraffa/issues/detail?id=30#c12) originally posted by shvachko on 2013-02-24T00:52:35.000Z:

After evaluating different options of which web server would be the best for giraffa, it turned out that reusing HttpServer infrastructure provided by hadoop-common (org.apache.hadoop.http.HttpServer) is the best way to go. HBase already provides all the required dependencies to run webserver, therefore giraffa will not need to import any additional dependencies. 

There's one challenge though: 
Giraffa webserver need to start as a MasterObserver, since it will run on HMaster.
Unfortunately it's not possible to load MasterObserver from the table descriptor as it's done for BlockManagementAgent and BaseEndpointCoprocessor in NamespaceAgent,
therefore as a part of giraffa installation hbase-site.xml will need to be modified, to define "hbase.coprocessor.master.classes" property, e.g. 

<property>
        <name>hbase.coprocessor.master.classes</name>
        <value>org.apache.giraffa.web.GiraffaWebObserver</value>
</property>

Except of above mentioned limitation everything else seems to be ok.

**Comment by shvachko** [comment]
Comment [#13](https://code.google.com/p/giraffa/issues/detail?id=30#c13) originally posted by shvachko on 2013-03-02T02:02:02.000Z:

Sounds like the right direction to add Giraffa web UI to HMaster via MasterObserver.
Do you plan to override postStartMaster()?

**Comment by shvachko** [comment]
Comment [#14](https://code.google.com/p/giraffa/issues/detail?id=30#c14) originally posted by shvachko on 2013-03-02T02:41:04.000Z:

Yes, that's exactly what I did. Web UI is already complete 70%. I'll try to get the rest done over the weekend and should be ready with patch in the beginning of the next week.

**Comment by shvachko** [comment]
Comment [#15](https://code.google.com/p/giraffa/issues/detail?id=30#c15) originally posted by shvachko on 2013-03-07T23:58:15.000Z:

Web UI patch is attached. It's 1.4 Mb because of new web resources (JS, css, img...)

apply this patch with 

git am --ignore-space-change --ignore-whitespace 0053-Issue-30-Added-Giraffa-Web-UI-Konstantin-P.patch

to test Web UI do "mvn clean install" followed by "mvn -Pwebdemo"

Demo Web Server will start on localhost:40010 and will contain some demo files.

to gracefully stop demo web server, type "stop" in console;

I did a lot of QA, everything seems to be working fine, but let me know if something is not working for you, or needs to be changed.

**Comment by shvachko** [comment]
Comment [#16](https://code.google.com/p/giraffa/issues/detail?id=30#c16) originally posted by shvachko on 2013-03-08T02:14:44.000Z:

Beautiful. We have reached our issue attachment storage quota with this issue though. :P Konst, is there any permanent deletes we can do for previous attachments?

Great work Kosta. I will give you some feedback later tonight. :)

**Comment by shvachko** [comment]
Comment [#17](https://code.google.com/p/giraffa/issues/detail?id=30#c17) originally posted by shvachko on 2013-03-08T02:48:58.000Z:

Okay there are a few things to go over actually. The work is all fine.

1) You add a "import org.apache.giraffa.*" in NamespaceProcessor.

2) I really like the use of NamespaceUtils, but can this be handled by a separate issue? I believe issue 51 may be more fitting for these changes.

3) #&nbsp;2 includes the addition of the FileField changes. I appreciate the change, but now we are sneaking in code changes if we commit this. Keep the logic changes in separate patches please.

4) There are many places where you break the 80 character line. Please be careful. Konst still prefers this limit; as do others.

5) Was the introduction of hbase-site.xml needed? It is empty.

6) There is a readme in the resources folder that explains your reasoning for putting webapps in the resource folder; that info would be better kept here I believe. Konst, if you think we should keep the readme than I am okay with it.

7) I am not sure if I agree with the RowKey API change. I am not convinced that we should be creating RowKey's primarily out of bytes. I think we can make a private exception for FullPathRowKey because the bytes == the path; however, I don't think this case can be made for generic RowKeys. Konst, thoughts?

8) You forgot to include your name in ()'s in the CHANGES.txt. :)

-Plamen

**Comment by shvachko** [comment]
Comment [#18](https://code.google.com/p/giraffa/issues/detail?id=30#c18) originally posted by shvachko on 2013-03-08T07:43:13.000Z:

1) my bad. IntelliJ Idea did this import automatically after it detects more than X imports from one package. 

2 & 3) all changes in code were introduced in order to get web functional. If it makes sense to move some of code changes into separate issues, let's do it. Since it's first open source project for me, I was not aware of best practices of code separation. 

4) Again my bad, again I was not aware of code style used in this project. By the way, are there any formatting or code guidelines?

5) hbase-site.xml was introduced only to demonstrate how web ui can be enabled in production environment. What's the best way to do that through documentation or sample xml file? 

7) Changes in RowKey api was introduced because I couldn't find anything to deserialize bytes into RowKey and it's used to read keys from Hbase to display in web page. Let me know if there's an existing API for this.

8) ops... :)

**Comment by shvachko** [comment]
Comment [#19](https://code.google.com/p/giraffa/issues/detail?id=30#c19) originally posted by shvachko on 2013-03-10T06:27:02.000Z:

Attached is an updated version of original patch:

https://s3.amazonaws.com/zapisnaya/public/issue30.patch

1)  jQuery, Datatables and Twitter Bbootstrap frameworks are now maven dependencies and downloaded during  build as per our email discussion. webapp still includes some css and js resources directly related to Giraffa webapp. It also includes minimal js and css files that were not available in maven repository).
Size of the patched got reduced to 212Kb as a result of these changes.

2) fixed #&nbsp;1

3) fixed #&nbsp;4

4) removed hbase-site.xml

5) removed new RowKey API. Moved it to FullPathRowKey

6) fixed #&nbsp;8

#&nbsp;2 and #&nbsp;3 are still not fixed. Once this patched is approved, I will move NamespaceUtils and FileField into issue 51.

**Comment by shvachko** [comment]
Comment [#20](https://code.google.com/p/giraffa/issues/detail?id=30#c20) originally posted by shvachko on 2013-03-13T07:31:31.000Z:

Sorry for delays in review. Lrge patches is harder to review. With dependencies it looks much more compact.
1. I think there are a few more things that need to be un-borrowed, if possible. Can you exclude
   /src/main/resources/hbase-webapps/giraffa/static/js/jquery.form.js
   and a couple of bootstrap .js files.
2. Kostya, you introduced grfa.web.bindAddress and grfa.web.port as different config parameters. Lets make it one as in HDFS, with the following key and default value:
   grfa.http-address=0.0.0.0:11110
   And you don't need to change the license comment. Probably DE did as well.
3. Saw some files .css without the Apache license at the top. Could you please scan new files for it.

I need to look more closely into RowKey changes.

We need more space for attachments. Plamen are you going to ask for it?

**Comment by shvachko** [comment]
Comment [#21](https://code.google.com/p/giraffa/issues/detail?id=30#c21) originally posted by shvachko on 2013-03-13T17:46:49.000Z:

Turned out that bootstrap-popover.js and bootstrap-tooltip.js are already bundled into main bootstrap.js which is a part of maven dependencies already, so it was safe to remove them from the project. 

Unfortunately with jquery-forms and bootstrap-fileupload it's going to be a bit harder, since these libraries are not available as maven artifacts yet. 

I think the best way to unborrow jquery-forms and bootstrap-fileupload resources would be to convert them to maven artifacts (just as I did for bootstrap, jquery and DataTables) and use them as maven dependencies. I'll contribute to http://www.webjars.org and add those artifacts to maven central as a part of webjars. When dependencies are available in maven central we would be able to use them in Giraffa. 

In this case we would keep only DT_bootstrap.js and filebrowser.js which are specific for Giraffa UI. 

What do you think? is it right way to go?

**Comment by shvachko** [comment]
Comment [#22](https://code.google.com/p/giraffa/issues/detail?id=30#c22) originally posted by shvachko on 2013-03-15T18:28:05.000Z:

Definitely. Sounds like a good plan.
I'll go over row key and file fields changes over the weekend.

**Comment by shvachko** [comment]
Comment [#23](https://code.google.com/p/giraffa/issues/detail?id=30#c23) originally posted by shvachko on 2013-03-18T00:08:38.000Z:

Three files from the patch are not applying to current trunk:
FileField.java
GiraffaConfiguration.java
NamespaceProcessor.java

**Comment by shvachko** [comment]
Comment [#24](https://code.google.com/p/giraffa/issues/detail?id=30#c24) originally posted by shvachko on 2013-03-18T01:43:08.000Z:

Applied with --ignore-whitespace. Some spaces are off.

**Comment by shvachko** [comment]
Comment [#25](https://code.google.com/p/giraffa/issues/detail?id=30#c25) originally posted by shvachko on 2013-03-18T02:24:50.000Z:

Changes to FileField doesn't make sense for two main reasons:
1. Package org.apache.giraffa should not have any dependencies on hbase. We moved all these dependencies under hbase package, see issues #&nbsp;12, #&nbsp;13, #&nbsp;17. The idea is to allow use of other databases to implement distributed namespace table.
2. Don't like using reflections for every call. It is not consistent with OOP and inefficient.
What problem were you trying to solve with these changes? Can you just use current FileField as enum?

**Comment by shvachko** [comment]
Comment [#26](https://code.google.com/p/giraffa/issues/detail?id=30#c26) originally posted by shvachko on 2013-03-18T08:39:21.000Z:

1) Brining hbase into org.apache.giraffa was my mistake, I should have checked #&nbsp;12,#&nbsp;13 and #&nbsp;17, my apologies. 

2) Main idea behind adding generics was that I didn't want to hard code FileType names in GiraffaHbaseServlet when going though result set, so that latter new FileTypes could be added and GiraffaHbaseServlet would not require any modification to it.  I did two major changes to FileTypes: 

2.a) I used generics in FileTypes to reduce code duplication, so code

for (FileField entry : FileField.values()) {
switch (entry) {
   case A_TIME:
     data.add(NamespaceProcessor.getATime(r)); break;
   case ACTION:
     data.add(new String(r.getValue(FileField.getFileAttributes(),
                    FileField.getAction()))); break;
   case NAME:
     data.add(NamespaceProcessor.getFileName(r));
     break;
   etc...
}}

could be replaced with: 

for (Map.Entry<String, FileField> entry : FileField.fieldMap.entrySet()) {
   data.add(entry.getResult(r));
}

2.b) I moved value extraction logic which used to be in static methods e.g. NamespaceProcessor: getBlocks(r), getFileAttributes(r), getFileName(r) into a function associated with each FileField, e.g.

public static Function<byte[], Long> TO_LONG = new Function<byte[], Long>() {
    public Long apply(byte[] bytes) {
      return Bytes.toLong(bytes);
    }
  };

FileField<Long> M_TIME = new FileField<Long>("mtime", TO_LONG);
FileField<Long> A_TIME = new FileField<Long>("atime", TO_LONG);
FileField<Short> REPLICATION = new FileField<Short>("replication", TO_SHORT);

so that code 

String value = new String(r.getValue(FileField.getFileAttributes(),FileField.getAction()))

can be rewritten as

String value = FileField.ACTION.getResult(r);

---

For now I'll revert back to previous version of FileField and will use static methods of NamespaceProcessor to extract values from Result object, if there will be a need to generalize FileTypes latter we can do it as a separate issue.

**Comment by shvachko** [comment]
Comment [#27](https://code.google.com/p/giraffa/issues/detail?id=30#c27) originally posted by shvachko on 2013-03-18T22:09:23.000Z:

Understand the reason for converting the enum FileField to generics now. Agreed, lets keep the old version and address it later when needed.

Here is the rest of the review comment. Sorry, taking longer than we all want.

RowKey.readKeyFromBytes()
- You should use RowKeyFactory.newInstance(src, bytes) rather than instantiating FullPathRowKey() directly. When we use another RowKey implementation this should work as well.
  I admit the API for RowKey is a bit tricky, and we need to clarify it in JavaDoc much more. Should be next item on our list. 

GiraffaWebJsonWrappers
- Unused fields: iColumns, bRegex, iTotalRecords, iTotalDisplayRecords, sEcho
- Unused fields in LocatedBlockDescriptor: all 5 of them. Do I understand correctly that you need LocatedBlockDescriptor class to attach those Json annotations? Can you just extend LocatedBlockDescriptor from LocatedBlock? If not you can fix unused field warnings by either removing "private" or by provding getters.
- Same with LocatedBlocksWrapper. There is LocatedBlocks class in HDFS.
- Also check the string values in this annotation
  @JsonPropertyOrder({"blockName", "size", "offset", "corrupt", "loc"})
  They probably should be the same as the field names, that is "locs" instead of "loc".
- FileItem has most fields from INode but named differently. Would it make sense to derive FileItem from INode and reuse the fields. Otherwise it is going to be a nightmare to support mapping of the two sets of fields in sync if they change in the future.

GiraffaFileServlet
- List items; List<E> should be parametrized
- In general, check for unused imports. I saw them in several new files.

Would be good to use s3 for the next iteration of the patch, since it is still rather large. We will attach the last committed version to the jira.

**Comment by shvachko** [comment]
Comment [#28](https://code.google.com/p/giraffa/issues/detail?id=30#c28) originally posted by shvachko on 2013-03-26T06:02:25.000Z:

I'm sorry for a long pause, but I'm just waiting for jquery-fileupload plugin to be added to webjars 

https://github.com/webjars/webjars/issues/91

As soon as it's in maven repo I'll post a final patch, which will address all issues that Konst mentioned in previous message.

**Comment by shvachko** [comment]
Comment [#29](https://code.google.com/p/giraffa/issues/detail?id=30#c29) originally posted by shvachko on 2013-03-26T07:36:27.000Z:

Great. Good to know.

**Comment by shvachko** [comment]
Comment [#30](https://code.google.com/p/giraffa/issues/detail?id=30#c30) originally posted by shvachko on 2013-04-02T17:44:29.000Z:

Just heard back from James Ward, who is managing webjars and he will his best to integrate jquery-fileupload to webjars and push it the central. Hope we will be able to wrap this up pretty soon.

**Comment by shvachko** [comment]
Comment [#31](https://code.google.com/p/giraffa/issues/detail?id=30#c31) originally posted by shvachko on 2013-04-13T00:00:05.000Z:

Unfortunately I'm not seeing any progress on 

https://github.com/webjars/webjars/issues/91

which is holding this issue from being resolved. Maybe we can include this one js file into Giraffa UI till webjars publishes it to maven central and we will just remove it from Giraffa UI when it gets to maven central?

**Comment by shvachko** [comment]
Comment [#32](https://code.google.com/p/giraffa/issues/detail?id=30#c32) originally posted by shvachko on 2013-04-13T01:00:31.000Z:

I would be okay with that. We have a plan to remove it when it comes time. Konst, this is also stalling us and it would be nice to move forward with other issues. I would also like to be able to host a hackathon soon for Giraffa. What do you think?

I'd say its fine Kosta. But should wait for Konst's response.

**Comment by shvachko** [comment]
Comment [#33](https://code.google.com/p/giraffa/issues/detail?id=30#c33) originally posted by shvachko on 2013-04-14T16:34:48.000Z:

Good news! just got this from James Ward: Sorry for the delay! This webjar is on it's way to Maven Central. If you have any questions or problems, file them at: https://github.com/webjars/jquery-form

So I should be wrapping this task up pretty soon

**Comment by shvachko** [comment]
Comment [#35](https://code.google.com/p/giraffa/issues/detail?id=30#c35) originally posted by shvachko on 2013-04-23T07:48:45.000Z:

Patch for issue 30 attached. It addresses all points that Konst brought up regarding GiraffaWebJsonWrappers, RowKey.readKeyFromBytes(), js/css resources.
All JS/CSS resources are included as Maven artifacts. Only relevant to Giraffa UI js/css files are distributed as a part of Giraffa webapp. 

To apply patch use 
git apply --ignore-space-change --ignore-whitespace issue30.patch

I tried all possible ways to get rid of whitespace problem, but it keeps showing up in CHANGES.txt file.

**Comment by shvachko** [comment]
Comment [#36](https://code.google.com/p/giraffa/issues/detail?id=30#c36) originally posted by shvachko on 2013-04-25T07:15:48.000Z:

I just tried this out. Looking good!

Things to address in later work:
1) Super stupid tiny nit: "Add a webui for Giraffa" is the title of issue in CHANGES.txt.
2) There are binary marks when displaying the rows in the HTable Browser? Looks very weird. We should make a separate issue where we should display this value as raw bytes instead of a String.
3) Currently we require webdemo profile to be active to run -- without the profile active most web classes will fail to compile; GiraffaFileServlet for example will fail to find org.apache.commons.\* imports. I think the web classes should always compile... whether or not we are demoing, no? We should be able to show off webui in single node Giraffa.

Other than that I give this a +1! Konst I want to commit this, may I? I've tweeted a screenshot of the UI running.

**Comment by shvachko** [comment]
Comment [#37](https://code.google.com/p/giraffa/issues/detail?id=30#c37) originally posted by shvachko on 2013-04-25T07:21:39.000Z:

Plamen, that's really strange that you can't compile without webdemo profile. This profile was not designed to add any extra dependencies for compilation and compile should work fine without webdemo. Before committing I would want to address this problem or at least verify that it's only specific to your environment. 

Can you please do "mvn clean install" and should me the error that you're seeing.

**Comment by shvachko** [comment]
Comment [#38](https://code.google.com/p/giraffa/issues/detail?id=30#c38) originally posted by shvachko on 2013-04-25T07:40:19.000Z:

The issue happened when I initially tried to run the demo in IntelliJ without the profile flag set, which caused those compilation errors. After mvn install I can run the demo WITHOUT the profile flag in IntelliJ and it runs fine.

Command line had the same thing.

Dependencies should have all been pulled on compilation, right? Weird that commons.fileupload was not part of it initially... I'll try to recreate it.

**Comment by shvachko** [comment]
Comment [#39](https://code.google.com/p/giraffa/issues/detail?id=30#c39) originally posted by shvachko on 2013-04-25T07:56:53.000Z:

Konstantin P, seems this was just in IntelliJ. Cannot reproduce it even after I wiped out .m2/ dir.

**Comment by shvachko** [comment]
Comment [#40](https://code.google.com/p/giraffa/issues/detail?id=30#c40) originally posted by shvachko on 2013-04-25T08:05:23.000Z:

yeah, most likely IntelliJ idea, I've seen it multiple times actually. Usually , dropping libraries cache in project settings helps

**Comment by shvachko** [comment]
Comment [#41](https://code.google.com/p/giraffa/issues/detail?id=30#c41) originally posted by shvachko on 2013-04-25T08:15:31.000Z:

I am unable to display the icons you mentioned offline. :( The binary mark goes away on different browsers.

The full byte key when using FullPathRowKey is: 1 byte (short) + fullpath bytes.

The short is the depth of the path (3 directories deep, 4 directories, etc), suffixed so that we maintain depth ordering in HBase. Therefore it is a number usually < 10, in our use cases anyway. When converted to a String this short value is some operational character that cannot be displayed properly.

**Comment by shvachko** [comment]
Comment [#42](https://code.google.com/p/giraffa/issues/detail?id=30#c42) originally posted by shvachko on 2013-04-25T08:24:30.000Z:

The line number causing this issue is line 134 on GiraffaHbaseServlet.java:

We can do: data.put(0, r.getRow());

This will display the byte-encoded String of the key (ex: AAEva29zdHlhJ3MgdGVzdA==). But now we don't see the full path anymore; however I think that is OK because people can use the HDFS browser anyway.

**Comment by shvachko** [comment]
Comment [#43](https://code.google.com/p/giraffa/issues/detail?id=30#c43) originally posted by shvachko on 2013-04-28T22:46:54.000Z:

Good job +1
Let's commit it and take care of the problems in subsequent issues.

pom.xml has some bogus pluses in <pluginManagement>

**Comment by shvachko** [comment]
Comment [#44](https://code.google.com/p/giraffa/issues/detail?id=30#c44) originally posted by shvachko on 2013-04-29T01:19:06.000Z:

Kostya, you should be able to commit this yourself.
I moved you to committers.

**Comment by shvachko** [comment]
Comment [#45](https://code.google.com/p/giraffa/issues/detail?id=30#c45) originally posted by shvachko on 2013-05-01T01:18:10.000Z:

Committed.
